### PR TITLE
Local MediaStreamTrack ended event

### PR DESCRIPTION
**Pull request by davidliu**
Addresses #1251.

Little unsure on what the spec specifically is, since at least [one browser seems to not fire the `ended` event upon user-initiated stop](https://stackoverflow.com/questions/55953038/why-is-the-ended-event-not-firing-for-this-mediastreamtrack); this implementation fires `ended` whenever the capturer is stopped, regardless of reason.

### CODE REVIEW

**Comment by 8BallBomBom** [comment]
Testing on Android seems good, will test on iOS soon.
As for the spec, unsure but will do some research 👍🏻

File: android/src/main/java/com/oney/WebRTCModule/ScreenCaptureController.java [hunk 1]
```diff
@@ -57,6 +57,7 @@ protected VideoCapturer createVideoCapturer() {
                 public void onStop() {
                     Log.w(TAG, "Media projection stopped.");
                     orientatationListener.disable();
+                    stopCapture();
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/ScreenCaptureController.java:60]: Is this change necessary because of this PR or was it necessary before?

File: android/src/main/java/com/oney/WebRTCModule/CameraEventsHandler.java [hunk 2]
```diff
@@ -13,7 +13,7 @@ class CameraEventsHandler implements CameraVideoCapturer.CameraEventsHandler {
     // Callback invoked when camera closed.
     @Override
     public void onCameraClosed() {
-        Log.d(TAG, "CameraEventsHandler.onFirstFrameAvailable");
+        Log.d(TAG, "CameraEventsHandler.onCameraClosed");
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/CameraEventsHandler.java:16]: Thanks for taking care of this!

File: android/src/main/java/com/oney/WebRTCModule/TrackCapturerObserver.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  /** Notify that the capturer has been stopped. */
+  public void onCapturerStopped() {
+    this.capturerObserver.onCapturerStopped();
+
+    WritableMap params = Arguments.createMap();
+    params.putString("trackId", trackId);
+
+    Log.d(TAG, "ended event trackId: " + trackId);
+
+    webRTCModule.sendEvent(
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/TrackCapturerObserver.java:44]: Technically a track ID can be duplicated across different peer connections. That's why we always have the PC ID to go with track events.

**Comment by saghul** [comment]
I'm a little torn on this one. According to Jan-Ivar, who works at Mozilla, it should not fire if the user stopped the track themselves: https://stackoverflow.com/a/55960232

That only leaves the abrupt terminations on the table, or more clearly stopping a screen capture with the system stop button.

File: android/src/main/java/com/oney/WebRTCModule/ScreenCaptureController.java (hunk 1 above)
Comment (davidliu) [reply android/src/main/java/com/oney/WebRTCModule/ScreenCaptureController.java:60]: Internally, `ScreenCapturerAndroid` doesn't listen to the media projection `onStop` callback, so the capturer stays "running" until directly stopped. The PR itself relies on stopping the capture to know when to emit the ended event, but aside from that, seems like a good idea to have it stop alongside the media projection stop to release any resources immediately.

File: android/src/main/java/com/oney/WebRTCModule/TrackCapturerObserver.java (hunk 3 above)
Comment (davidliu) [reply android/src/main/java/com/oney/WebRTCModule/TrackCapturerObserver.java:44]: Hmm, this one is a little hairy since tracks are created independently from their eventual peer connection. I was figuring it'd be safe enough since we use UUIDs for the tracks. This'll need a little more thinking to insert the PC id into here when they get associated to each other.

Comment (saghul) [reply android/src/main/java/com/oney/WebRTCModule/TrackCapturerObserver.java:44]: I'll double check, since we have changed things around indeed.

File: android/src/main/java/com/oney/WebRTCModule/ScreenCaptureController.java (hunk 1 above)
Comment (saghul) [reply android/src/main/java/com/oney/WebRTCModule/ScreenCaptureController.java:60]: Can you pl send a PR with this change alone?

**Comment by davidliu** [comment]
@saghul Hmm, if that's the case, I think I can just add a flag to the capturers so as not emit the event when the sdk/user calls stop manually, does that sound reasonable?

**Comment by saghul** [comment]
I'd like to thread lightly here. My understanding is a track cannot be resumed once it gets to "ended".

**Comment by davidliu** [comment]
Hmm, is the concern that the user can somehow resume the capturer even after it's stopped? Not quite understanding the issue here.

**Comment by saghul** [comment]
> Hmm, is the concern that the user can somehow resume the capturer even after it's stopped? Not quite understanding the issue here.

Maybe I misunderstood what you meant, but if we emit ended when the capturer is interrupted, what happens when it resumes?

**Comment by davidliu** [comment]
AFAICT, the only capturer at the moment that resumes itself is iOS's camera capturer (this one still needs to be addressed, or left as a TODO). The other ones don't have any sort of resume functionality and stop indefinitely until `startCapture` is called again.

Normally this would fine since we the web API of MediaStreamTrack doesn't have start function, but did find a snag here; `MediaStreamTrack.setEnabled` additionally starts and stops the capturer, so users could resume a capturer that way.

Checked on Chrome and Safari, they seem to keep the capturer going while the track is disabled, while Firefox stops the capturer while the track is disabled. If we went the way of Chrome, that could solve this issue though might be a big unexpected change for users of the library.

**Comment by saghul** [comment]
> AFAICT, the only capturer at the moment that resumes itself is iOS's camera capturer (this one still needs to be addressed, or left as a TODO). The other ones don't have any sort of resume functionality and stop indefinitely until `startCapture` is called again.

Right.



> Normally this would fine since we the web API of MediaStreamTrack doesn't have start function, but did find a snag here; `MediaStreamTrack.setEnabled` additionally starts and stops the capturer, so users could resume a capturer that way.

That's an internal API and it's not exposed so that's ok.



> Checked on Chrome and Safari, they seem to keep the capturer going while the track is disabled, while Firefox stops the capturer while the track is disabled. If we went the way of Chrome, that could solve this issue though might be a big unexpected change for users of this library.

I implemented the Firefox behavior explicitly :-) Google once told me they'd eventually implement it too but was not high priority. Stopping the capture when enabled is set to false makes video muting way faster because there is no need to destroy and re-create the track, as we have to do on Chrome, for example.

**Comment by davidliu** [comment]
>>Normally this would fine since we the web API of MediaStreamTrack doesn't have start function, but did find a snag here; MediaStreamTrack.setEnabled additionally starts and stops the capturer, so users could resume a capturer that way.

>That's an internal API and it's not exposed so that's ok.

I mention it here because although it's internal behavior, it's something a user might inadvertently do; it would go against spec if we emit an "ended" event due to camera error, but the user flips the enabled event and winds up successfully restarting the capturer.

>I implemented the Firefox behavior explicitly :-) Google once told me they'd eventually implement it too but was not high priority. Stopping the capture when enabled is set to false makes video muting way faster because there is no need to destroy and re-create the track, as we have to do on Chrome, for example.

Hmm, if that's the case, I think I'll abandon trying to do `ended` events for the camera cases since there's no permanent end state that I know of there (aside from permission revocation, but those kill the app on iOS/Android), and just leave the ended events for the external screen capture stop.

**Comment by saghul** [comment]
Sounds good to me!

**Comment by davidliu** [comment]
Cleaned up the PR to only do screen capturer events, take a look when you can.

File: src/MediaStreamTrack.ts [hunk 4]
```diff
@@ -36,6 +39,8 @@ class MediaStreamTrack extends defineCustomEventTarget(...MEDIA_STREAM_TRACK_EVE
         this.id = info.id;
         this.kind = info.kind;
         this.remote = info.remote;
+
+        this._registerEvents();
```
Comment (saghul) [inline src/MediaStreamTrack.ts:43]: We should only do this for local tracks. It's not necessary to release remote tracks, so those will leak.

**Review by saghul — CHANGES_REQUESTED** [review]
One final comment, LGTM otherwise.

File: src/MediaStreamTrack.ts (hunk 4 above)
Comment (davidliu) [reply src/MediaStreamTrack.ts]: Added in a check on `this.remote` and tested, seems to work fine there. 

A little worried about the SerializeUtils functions always returning true for remote though:

https://github.com/react-native-webrtc/react-native-webrtc/blob/master/android/src/main/java/com/oney/WebRTCModule/SerializeUtils.java#L76

https://github.com/react-native-webrtc/react-native-webrtc/blob/0651e2b0600435dacd7b1e13d5f47914252e1e35/ios/RCTWebRTC/SerializeUtils.m#L174

I don't think there's a way for this flag to be incorrectly set for local tracks atm though, so should be fine if this isn't misused.

Comment (saghul) [reply src/MediaStreamTrack.ts]: Hum good point, we should check that.

**Review by saghul — APPROVED** [review]
